// react-native-simple-twitter/dist/util.js
import HmacSHA1 from 'crypto-js/hmac-sha1';
import * as Base64 from 'crypto-js/enc-base64';
export const randomStrings = (n = 32) => {
    const str = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    return Array(...Array(n)).map(() => str.charAt(Math.floor(Math.random() * str.length))).join('');
};
export const createHeaderString = (params) => `OAuth ${Object.keys(params).sort()
    .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
    .join(', ')}`;
export const encodeParamsToString = (params) => Object.keys(params).sort()
    .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
    .join('&');
export const parseFormEncoding = (formEncoded) => formEncoded.split('&').reduce((obj, form) => {
    const [key, value] = form.split('=');
    return { ...obj, [key]: value };
}, {});
export const createTokenRequestHeaderParams = (consumerKey, { callback, token, params }) => ({
    ...(callback ? { oauth_callback: callback } : {}),
    oauth_consumer_key: consumerKey,
    oauth_nonce: randomStrings(),
    oauth_signature_method: 'HMAC-SHA1',
    oauth_timestamp: new Date().getTime() / 1000,
    ...(token ? { oauth_token: token } : {}),
    oauth_version: '1.0',
    ...params,
});
export const createSignature = (params, method, url, consumerSecret, tokenSecret) => {
    const encodedParameters = encodeParamsToString(params)
        .replace(/!/g, '%21')
        .replace(/'/g, '%27')
        .replace(/\(/g, '%28')
        .replace(/\)/g, '%29')
        .replace(/\*/g, '%2A');
    const encodedRequestURL = encodeURIComponent(url);
    const signature = Base64.stringify(HmacSHA1(`${method}&${encodedRequestURL}&${encodeURIComponent(encodedParameters)}`, tokenSecret ? `${encodeURIComponent(consumerSecret)}&${encodeURIComponent(tokenSecret)}` : `${encodeURIComponent(consumerSecret)}&`));
    return { ...params, oauth_signature: signature };
};
